File: python/matching-brackets/matching_brackets.py
CLOSING = {"{": "}", "[": "]", "(": ")"}


def is_paired(input_string):
    stack = []
    for c in filter(lambda c: c in "{[()]}", input_string):
        if c in "{[(":
            stack.append(c)
        elif not stack or CLOSING[stack.pop()] != c:
            return False
    return not stack

File: ruby/matching-brackets/matching_brackets.rb
module Brackets
  CLOSING = {"(" => ")", "[" => "]", "{" => "}"}
  def self.paired?(string)
    stack = []
    string.chars().filter {|c| "{[()]}".include?(c) }.each do |c|
      if "{[(".include?(c) then 
        stack.push(c)
      elsif stack.empty? or CLOSING[stack.pop()] != c then
        return false
      end
    end
    stack.empty?
  end
end
